GitBitterLib/Implementations/GitSharpCloner.cs

```diff
@@ -123,27 +131,68 @@ public Task ResetAndUpdateExisting(string repository, string rootdir, string rep
+        private Branch GetOrCreateLocalBranch(IRepository repo, string branchname)
         {
-            var filesAndFolders = GitBitterContainer.Default.Resolve<IGitFilesAndFolders>();
+            var localBranch = repo.Branches[branchname];
 
-            var credentials = new SshUserKeyCredentials();
-            credentials.Username = "git";
-            credentials.Passphrase = string.Empty;
-            credentials.PrivateKey = filesAndFolders.SSHPrivateKeyFile();
-            credentials.PublicKey = filesAndFolders.SSHPublicKeyFile();
+            if (localBranch == null)
+            {
+                var remoteBranch = GetRemoteBranch(repo, branchname);
+                if (remoteBranch != null)
+                {
+                    return CreateLocalBranchFromRemote(repo, remoteBranch, branchname);
+                }
+            }
+
+            return localBranch;
+        }
 
-            if (!File.Exists(credentials.PrivateKey))
+        private Branch CreateLocalBranchFromRemote(IRepository repo, Branch remotebranch, string branchname)
+        {
+            var localbranch = repo.CreateBranch(branchname, remotebranch.Tip);
+
+            repo.Branches.Update(localbranch, b => b.TrackedBranch = remotebranch.CanonicalName);
+
+            return localbranch;
+        }
+
+        private Branch GetRemoteBranch(IRepository repo, string branchname)
+        {
+            foreach (var branch in repo.Branches)
             {
-                throw new FileNotFoundException("Private key not found @ " + credentials.PrivateKey);
+                if ((branch.FriendlyName == branch.RemoteName + "/" + branchname))
+                {
+                    return branch;
+                }
             }
 
-            if (!File.Exists(credentials.PublicKey))
+            return null;
+        }
+
+        private CredentialsHandler GetCredentialHandlerSSH(string repository)
+        {
+            var filesAndFolders = GitBitterContainer.Default.Resolve<IGitFilesAndFolders>();
+
+            var privkey = filesAndFolders.SSHPrivateKeyFile();
+            var pubkey = filesAndFolders.SSHPublicKeyFile();
+
+            if (!File.Exists(privkey) || !File.Exists(pubkey))
             {
-                throw new FileNotFoundException("Public key not found @ " + credentials.PublicKey);
+                var credentials = new SshAgentCredentials();
+                credentials.Username = "git";
+
+                return (_url, _user, _cred) => credentials;
             }
+            else
+            {
+                var credentials = new SshUserKeyCredentials();
+                credentials.Username = "git";
+                credentials.Passphrase = string.Empty;
+                credentials.PrivateKey = privkey;
+                credentials.PublicKey = pubkey;
 
-            return (_url, _user, _cred) => credentials;
+                return (_url, _user, _cred) => credentials;
```

Comment:
yes, need to figure out what the base class is of both credential objects though.. but, will do